cli/internal/runsummary/execution_summary.go

```diff
@@ -0,0 +1,201 @@
+	// Its current status
+	Status executionEventName
+	// Error, only populated for failure statuses
+	Err error
+}
+
+// executionEventName represents the status of a target when we log a build result.
+type executionEventName int
+
+// The collection of expected build result statuses.
+const (
+	targetBuilding executionEventName = iota
+	TargetBuildStopped
+	TargetBuilt
+	TargetCached
+	TargetBuildFailed
+)
+
+func (en executionEventName) toString() string {
+	switch en {
+	case targetBuilding:
+		return "building"
+	case TargetBuildStopped:
+		return "buildStopped"
+	case TargetBuilt:
+		return "built"
+	case TargetCached:
+		return "cached"
+	case TargetBuildFailed:
+		return "buildFailed"
+	}
+
+	return ""
+}
+
+// TaskExecutionSummary contains data about the state of a single task in a turbo run.
+// Some fields are updated over time as the task prepares to execute and finishes execution.
+type TaskExecutionSummary struct {
+	StartAt time.Time `json:"start"`
+
+	Duration time.Duration `json:"duration"`
+
+	// Target which has just changed
+	Label string `json:"-"`
+
+	// Its current status
+	Status string `json:"status"`
+
+	// Error, only populated for failure statuses
+	Err error `json:"error"`
+}
+
+// executionSummary is the state of the entire `turbo run`. Individual task state in `Tasks` field
+// TODO(mehulkar): Can this be combined with the RunSummary?
+type executionSummary struct {
+	mu        sync.Mutex                       `json:"-"`
+	state     map[string]*TaskExecutionSummary `json:"-"`
+	Success   int                              `json:"success"`
+	Failure   int                              `json:"failed"`
+	Cached    int                              `json:"cached"`
+	Attempted int                              `json:"attempted"`
+
+	startedAt time.Time
+
+	profileFilename string
+}
+
+// newExecutionSummary creates a executionSummary instance to track events in a `turbo run`.`
+func newExecutionSummary(start time.Time, tracingProfile string) *executionSummary {
+	if tracingProfile != "" {
+		chrometracing.EnableTracing()
+	}
+
+	return &executionSummary{
+		Success:         0,
+		Failure:         0,
+		Cached:          0,
+		Attempted:       0,
+		state:           make(map[string]*TaskExecutionSummary),
+		startedAt:       start,
+		profileFilename: tracingProfile,
+	}
+}
+
+// Run starts the Execution of a single task. It returns a function that can
+// be used to update the state of a given taskID with the executionEventName enum
+func (es *executionSummary) run(label string) (func(outcome executionEventName, err error), *TaskExecutionSummary) {
+	start := time.Now()
+	taskExecutionSummary := es.add(&executionEvent{
+		Time:   start,
+		Label:  label,
+		Status: targetBuilding,
+	})
+
+	tracer := chrometracing.Event(label)
+
+	// This function can be called with an enum and an optional error to update
+	// the state of a given taskID.
+	tracerFn := func(outcome executionEventName, err error) {
+		defer tracer.Done()
```

Comment:
Does this mean that the tracerFn is meant to only be called once? Is it perhaps possible to track and return an error if you call it twice? Otherwise add will register the same event twice (and increment Failure, Attempted, etc)

Reply:
It's confusingly named, but `tracer` and `tracerFn` are different things. 
- `tracer` is for the _chrometracing_ profile that's generated
- `tracerFn` updates the internal status

this `tracer.Done()` is for chrometracing profile. I'm not entirely sure how it works, but I tested before and after this PR and the generated `--profile` is the same